### DISCUSSION TITLE
Difficulty testing my trained model

### DISCUSSION THREAD

**Discussion by Biribbissolo** [open-mmlab/mmclassification#1111] [General]
Greetings,
I used the script made available on colab for a complete training, backbone and head with the dataset of dogs and cats. (https://colab.research.google.com/github/open-mmlab/mmclassification/blob/master/docs/en/tutorials/MMClassification_python.ipynb)
Here are the scripts I used for cfg and training ( it worked great ):

```
#Load the base config file
from mmcv import Config
cfg = Config.fromfile('configs/mobilenet_v2/mobilenet-v2_8xb32_in1k.py')

#Modify the number of classes in the head.
cfg.model.head.num_classes = 2                     # Cats and Dogs
cfg.model.head.topk = (1, )

cfg.device='cuda'                      
cfg.model.backbone.frozen_stages = -1    # Set all layers trainable

#Specify sample size and number of workers.
cfg.data.samples_per_gpu = 32
cfg.data.workers_per_gpu = 2

#Specify the path and meta files of training dataset
cfg.data.train.data_prefix = 'data/cats_dogs_dataset/training_set/training_set'
cfg.data.train.classes = 'data/cats_dogs_dataset/classes.txt'

#Specify the path and meta files of validation dataset
cfg.data.val.data_prefix = 'data/cats_dogs_dataset/val_set/val_set'
cfg.data.val.ann_file = 'data/cats_dogs_dataset/val.txt'
cfg.data.val.classes = 'data/cats_dogs_dataset/classes.txt'

#Specify the path and meta files of test dataset
cfg.data.test.data_prefix = 'data/cats_dogs_dataset/test_set/test_set'
cfg.data.test.ann_file = 'data/cats_dogs_dataset/test.txt'
cfg.data.test.classes = 'data/cats_dogs_dataset/classes.txt'

#Specify the normalization parameters in data pipeline
normalize_cfg = dict(type='Normalize', mean=[124.508, 116.050, 106.438], std=[58.577, 57.310, 57.437], to_rgb=True)
cfg.data.train.pipeline[3] = normalize_cfg
cfg.data.val.pipeline[3] = normalize_cfg
cfg.data.test.pipeline[3] = normalize_cfg

#Modify the evaluation metric
cfg.evaluation['metric_options']={'topk': (1, )}

#Specify the optimizer
cfg.optimizer = dict(type='SGD', lr=0.005, momentum=0.9, weight_decay=0.0001)
cfg.optimizer_config = dict(grad_clip=None)

#Specify the learning rate scheduler
cfg.lr_config = dict(policy='step', step=20, gamma=0.1)
cfg.runner = dict(type='EpochBasedRunner', max_epochs=50)

#Specify the work directory
cfg.work_dir = '/content/drive/MyDrive/GAN_FILTRATO/Classificatore_Finale/zona_lavoro'

#Output logs for every 10 iterations
cfg.log_config.interval = 100

#Set the random seed and enable the deterministic option of cuDNN
#to keep the results' reproducible.
from mmcls.apis import set_random_seed
cfg.seed = 0
set_random_seed(0, deterministic=True)

cfg.gpu_ids = range(1)



import time
import mmcv
import os.path as osp

from mmcls.datasets import build_dataset
from mmcls.models import build_classifier
from mmcls.apis import train_model

#Create the work directory
mmcv.mkdir_or_exist(osp.abspath(cfg.work_dir))
#Build the classifier
model = build_classifier(cfg.model)
model.init_weights()
#Build the dataset
datasets = [build_dataset(cfg.data.train)]
#Add `CLASSES` attributes to help visualization
model.CLASSES = datasets[0].CLASSES
#Start Trainng
train_model(
    model,
    datasets,
    cfg,
    distributed=False,
    validate=True,
    timestamp=time.strftime('%Y%m%d_%H%M%S', time.localtime()),
    meta=dict())
```

Once I finished training the 50 epochs, I tried to recreate the same configuration, and then load the pre-trained weights (both the backbone and the head), so that I could use the classifier, however, something went wrong in the phase of the loading the trained model. The net comes up with a result adapted to my case (cats and dogs), but it seems that the result is random, as if the weights were not loaded correctly.
Below is the code that recreates the cfg and loads the trained model:

```
import time
import mmcv
import os.path as osp
import time
import mmcv
import os.path as osp

from mmcls.datasets import build_dataset
from mmcls.models import build_classifier
from mmcls.apis import train_model

# Specify the path of the config file and checkpoint file.
config_file = 'configs/mobilenet_v2/mobilenet-v2_8xb32_in1k.py'
checkpoint_file = '/content/drive/MyDrive/GAN_FILTRATO/Classificatore_Finale/zona_lavoro/epoch_50.pth'

# Load the base config file
from mmcv import Config
cfg = Config.fromfile('configs/mobilenet_v2/mobilenet-v2_8xb32_in1k.py')

# Modify the number of classes in the head.
cfg.model.head.num_classes = 2
cfg.model.head.topk = (1, )

# Load the pre-trained model's checkpoint.
cfg.model.init_cfg = dict(type='Pretrained', checkpoint=checkpoint_file)

# Specify sample size and number of workers.
cfg.data.samples_per_gpu = 32
cfg.data.workers_per_gpu = 8

# Specify the path and meta files of training dataset
cfg.data.train.data_prefix = 'data/cats_dogs_dataset/training_set/training_set'
cfg.data.train.classes = 'data/cats_dogs_dataset/classes.txt'

# Specify the path and meta files of validation dataset
cfg.data.val.data_prefix = 'data/cats_dogs_dataset/val_set/val_set'
cfg.data.val.ann_file = 'data/cats_dogs_dataset/val.txt'
cfg.data.val.classes = 'data/cats_dogs_dataset/classes.txt'

# Specify the path and meta files of test dataset
cfg.data.test.data_prefix = 'data/cats_dogs_dataset/test_set/test_set'
cfg.data.test.ann_file = 'data/cats_dogs_dataset/test.txt'
cfg.data.test.classes = 'data/cats_dogs_dataset/classes.txt'


# Modify the evaluation metric
cfg.evaluation['metric_options']={'topk': (1, )}

# Specify the optimizer
cfg.optimizer = dict(type='SGD', lr=0.005, momentum=0.9, weight_decay=0.0001)
cfg.optimizer_config = dict(grad_clip=None)

# Specify the learning rate scheduler
cfg.lr_config = dict(policy='step', step=20, gamma=0.1)
cfg.runner = dict(type='EpochBasedRunner', max_epochs=50)

# Specify the work directory
cfg.work_dir = '/content/drive/MyDrive/GAN_FILTRATO/Classificatore_Finale/zona_lavoro'
# Output logs for every 10 iterations
cfg.log_config.interval = 100
cfg.device = 'cuda'

# Set the random seed and enable the deterministic option of cuDNN
# to keep the results' reproducible.
from mmcls.apis import set_random_seed
cfg.seed = 0
set_random_seed(0, deterministic=True)

cfg.gpu_ids = range(1)

# Create the work directory
mmcv.mkdir_or_exist(osp.abspath(cfg.work_dir))
# Build the classifier
model = build_classifier(cfg.model)
model.init_weights()
# Build the dataset
datasets = [build_dataset(cfg.data.train)]
# Add `CLASSES` attributes to help visualization
model.CLASSES = datasets[0].CLASSES
```
Sorry if I was redundant.
Finally, I tried to test the model like this. What am I wrong in the loading phase of the trained model? Thanks so much for your attention.

```
%matplotlib inline
# Validate the trained model
img = mmcv.imread('data/cats_dogs_dataset/training_set/training_set/cats/cat.1.jpg')
model.cfg = cfg
result = inference_model(model, img)
show_result_pyplot(model, img, result)
```
![gatto](https://user-images.githubusercontent.com/104684808/191361824-f5161aa6-a126-442a-83cf-05556d345aab.png)

**Comment by Ezra-Yu** [comment]
What is the accuracy of training weights for 50 epochs？ which would print in the training log.

**Comment by Biribbissolo** [comment]
I don't actually remember exactly because I lost the output, but I'm sure it was above 80%. If you want I can run it again, keeping the exit.

**Comment by Ezra-Yu** [comment]
what about using [`this demo`](https://github.com/open-mmlab/mmclassification/blob/master/demo/image_demo.py) to test?

Or try following:
```
import mmcv
from mmcls.apis import inference_model, init_model, show_result_pyplot

config = xxxx # copy and paste your model config
checkpoint_path = xxx
device = "cpu"
img_path = xxxx


model = init_model(config, checkpoint, device)
# test a single image
result = inference_model(model, img)
# show the results
print(mmcv.dump(result, file_format='json', indent=4))
# show the result
show_result_pyplot(model, args.img, result)
```

**Comment by Biribbissolo** [comment]
Thank you very much. It worked. 
The following link was very helpful to me: https://colab.research.google.com/github/open-mmlab/mmclassification/blob/master/docs/en/tutorials/MMClassification_tools.ipynb
